### PR TITLE
improvement/project-accessible-in-issue

### PR DESCRIPTION
**Pull request by MarcoGiesen**

### CODE REVIEW

**Comment by tolry** [comment]
Great, that you picked up the task, lazy loading is really the way to go here. Some notes:
- injecting the lazy loading closure, containing a repository, into an entity is quite ... problematic
- there is a kind of hidden contract there now: you have to use this setter-method to get a fully functional object
- I would suggest using a proxy object to wrap the entity and keep the entity very plain and dependency free - best lib for this is https://github.com/Ocramius/ProxyManager
- the process of creating the object from your JSON data, including the proxy wrapping, is currently duplicated Code in the repository. Extract it into a method or even a seperate class. This process is usually called hydration (e.g. in doctrine), if you are looking for a suitable name

**Comment by DavidBadura** [comment]
Hey, 

you have create a proxy from the wrong class. You should create it from Project instead Issue. You don't want to lazy load the issue which you get already. You want to load the Project if you access an attribute which aren't loaded.

``` php
$issue = $client->getIssueRepository()->find(1); // real object
$project = $issue->getProject(); // proxy from project (not initialized)
echo $project->getId(); // print 1 (not initialized)
echo $project->getName(); // print Test (initialize!)
```

Links:
- https://github.com/DavidBadura/Taskwarrior/blob/master/src/TaskManager.php#L310
- https://github.com/DavidBadura/Taskwarrior/blob/master/src/Proxy/UuidContainer.php

File: src/Repository/IssueRepository.php [hunk 1]
```diff
@@ -41,11 +46,16 @@ public function findAll()
     {
         $result = $this->client->get('issues.json?include=children,attachments,relations,changesets,watchers');
 
-        if (! $result) {
+        if(! $result) {
```
Comment (DanielBadura) [inline src/Repository/IssueRepository.php]: Why did you remove the whitespace?
PSR-2: http://www.php-fig.org/psr/psr-2/#1-1-example

File: src/Hydration/IssueProjectHydration.php [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        $initializer = function (
+            &$wrappedObject,
+            LazyLoadingInterface $proxy,
+            $method
+        ) use ($project, $projectRepository) {
+            if($method == "getIdentifier" ||
+                $method == "getDescription" ||
+                $method == "getCreatedOn" ||
+                $method == "getUpdatedOn" ||
+                $method == "isIsPublic" ||
```
Comment (DanielBadura) [inline src/Hydration/IssueProjectHydration.php]: isIsPublic? I think you meant isPublic

File: src/Hydration/IssueProjectHydration.php [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+ * @author Marco Giesen <marco.giesen93@gmail.com>
+ */
+class IssueProjectHydration
+{
+    /**
+     * @param Issue $issue
+     * @param Client $client
+     * @return Issue
+     */
+    public function hydrateOneIssue(Issue $issue, Client $client)
```
Comment (DanielBadura) [inline src/Hydration/IssueProjectHydration.php]: Naming it just `hydrateIssue(..)` should be enough in my opinion.
The other method for multiple issues can you leave like it is.

**Comment by DanielBadura** [comment]
Is this still WIP besides my comments?

File: src/Repository/IssueRepository.php (hunk 1 above)
Comment (MarcoGiesen) [inline src/Repository/IssueRepository.php]: We ignore the PSR-2 in every if statement at this point. Maybe we should reformat the code in another branch.

File: src/Hydration/IssueProjectHydration.php (hunk 2 above)
Comment (MarcoGiesen) [inline src/Hydration/IssueProjectHydration.php]: auto generated code, the value itself is "isPublic", named by redmine. I will rename it to simply "isPublic"

File: src/Hydration/IssueProjectHydration.php [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        $project = $issue->getProject();
+        $projectRepository = $client->getProjectRepository();
+
+        $factory = new LazyLoadingValueHolderFactory();
+        $initializer = function (
+            &$wrappedObject,
+            LazyLoadingInterface $proxy,
+            $method
+        ) use ($project, $projectRepository) {
+            if($method == "getIdentifier" ||
```
Comment (DavidBadura) [inline src/Hydration/IssueProjectHydration.php]: you should negate the statement

``` php
if ($method == 'getId') {
  $wrappedObject = $project;
} else {
  if(! $wrappedObject) {
    $wrappedObject = $projectRepository->find($project->getId());
    $proxy->setProxyInitializer(null);
  }
}
```

**Comment by DavidBadura** [comment]
another few points:
- you should implement an IdentityMap to ensure that there is only one instance of an entity.
- the hydration should happen in the serializer

**Comment by tolry** [comment]
basically copy doctrine's UOW :smile:
